File: zig/plugins/infisical.zig
//! Infisical - a sekreto plugin. Needs HTTPS.

const std = @import("std");

const sekreto = @import("sekreto");
const httpjson = @import("httpjson.zig");

const Allocator = std.mem.Allocator;
const Answer = sekreto.Answer;
const Found = sekreto.Found;
const Provider = sekreto.Provider;
const ProviderSpec = sekreto.ProviderSpec;

/// Infisical.
///
/// `api.token` reads the secret keyed `API_TOKEN` (Infisical's own
/// convention is environment-style keys) at a secret path in one environment
/// of a project. sekreto.Auth is a token, or a universal-auth (machine identity)
/// login with clientid/clientsecret.
pub const InfisicalProvider = struct {
    alloc: Allocator,
    config: sekreto.Config,
    addr: []const u8,
    token: []const u8,
    clientid: []const u8,
    clientsecret: []const u8,
    project: []const u8,
    environment: []const u8,
    path: []const u8,
    livetoken: ?[]const u8,
    renewat: i64,

    fn login(self: *InfisicalProvider, alloc: Allocator, addr: []const u8) Allocator.Error!Answer([]const u8) {
        if (0 != self.token.len) {
            return .{ .ok = self.token };
        }

        if (0 == self.clientid.len or 0 == self.clientsecret.len) {
            return .{ .err = try sekreto.fail(
                alloc,
                "sekreto: infisical: no token and no client credentials",
                .{},
            ) };
        }

        const url = try std.fmt.allocPrint(alloc, "{s}/api/v1/auth/universal-auth/login", .{addr});

        const body = try httpjson.jsonobject(alloc, &.{
            .{ .key = "clientId", .value = self.clientid },
            .{ .key = "clientSecret", .value = self.clientsecret },
        });

        const headers = [_]httpjson.Header{
            .{ .name = "content-type", .value = "application/json" },
        };

        const res = switch (try httpjson.fetchjson(alloc, self.config.io, .POST, url, &headers, body)) {
            .err => |message| return .{ .err = message },
            .ok => |got| got,
        };

        const access = httpjson.jstr(httpjson.jget(res.body, "accessToken"));

        if (200 != res.status or null == access) {
            return .{ .err = try sekreto.fail(
                alloc,
                "sekreto: infisical login failed: {d}",
                .{res.status},
            ) };
        }

        self.renewat = httpjson.renewafter(self.config.io, httpjson.jnum(httpjson.jget(res.body, "expiresIn")));

        return .{ .ok = access.? };
    }

    pub fn lookup(self: *InfisicalProvider, alloc: Allocator, name: []const u8) Allocator.Error!Found {
        const addr = httpjson.trimslash(if (0 != self.addr.len) self.addr else "https://app.infisical.com");

        switch (try sekreto.checkaddr(alloc, addr)) {
            .err => |message| return .{ .err = message },
            .ok => {},
        }

        if (0 == self.project.len or 0 == self.environment.len) {
            return .{ .err = try sekreto.fail(alloc, "sekreto: infisical: no project/environment", .{}) };
        }

        if (null == self.livetoken or httpjson.nowms(self.config.io) >= self.renewat) {
            const fresh = switch (try self.login(alloc, addr)) {
                .err => |message| return .{ .err = message },
                .ok => |got| got,
            };
            if (self.livetoken) |old| {
                self.alloc.free(old);
            }
            self.livetoken = try self.alloc.dupe(u8, fresh);
        }

        const key = switch (try sekreto.envkey(alloc, name, "")) {
            .err => |message| return .{ .err = message },
            .ok => |made| made,
        };

        const url = try std.fmt.allocPrint(
            alloc,
            "{s}/api/v3/secrets/raw/{s}?workspaceId={s}&environment={s}&secretPath={s}",
            .{
                addr,
                key,
                try httpjson.escape(alloc, self.project),
                try httpjson.escape(alloc, self.environment),
                try httpjson.escape(alloc, if (0 != self.path.len) self.path else "/"),
            },
        );

        const bearer = try std.fmt.allocPrint(alloc, "Bearer {s}", .{self.livetoken.?});
        const headers = [_]httpjson.Header{.{ .name = "authorization", .value = bearer }};

        const res = switch (try httpjson.fetchjson(alloc, self.config.io, .GET, url, &headers, null)) {
            .err => |message| return .{ .err = message },
            .ok => |got| got,
        };

        if (404 == res.status) {
            return .{ .ok = null };
        }

        if (200 != res.status) {
            return .{ .err = try sekreto.fail(alloc, "sekreto: infisical error: {d}", .{res.status}) };
        }

        return .{ .ok = httpjson.jstr(httpjson.jget(httpjson.jget(res.body, "secret"), "secretValue")) };
    }

    pub fn describe(self: *InfisicalProvider, alloc: Allocator) Allocator.Error![]const u8 {
        return std.fmt.allocPrint(alloc, "infisical:{s}/{s}", .{ self.project, self.environment });
    }

    pub fn deinit(self: *InfisicalProvider, alloc: Allocator) void {
        if (self.livetoken) |token| {
            alloc.free(token);
        }
    }
};

fn make(alloc: Allocator, config: sekreto.Config, spec: ProviderSpec) Allocator.Error!Answer(Provider) {
    return .{ .ok = try sekreto.provide(alloc, InfisicalProvider, .{
        .alloc = alloc,
        .config = config,
        .addr = spec.addr,
        .token = spec.token,
        .clientid = spec.clientid,
        .clientsecret = spec.clientsecret,
        .project = spec.project,
        .environment = spec.environment,
        .path = spec.path,
        .livetoken = null,
        .renewat = httpjson.NEVER,
    }) };
}

/// The `infisical` kind.
pub const infisical: sekreto.Definition = sekreto.providerplugin("infisical", make);

File: typescript/plugins/infisical.ts
/* Copyright (c) 2025 Voxgig Ltd, MIT License */

import { ProviderSpec, Provider, SekretoError, envkey, providerplugin } from '../src/provider/support'
import { checkaddr } from '../src/provider/addr'
import { fetchjson } from './httpjson'

/** Infisical.
 *
 * `api.token` reads the secret keyed `API_TOKEN` (Infisical's own
 * convention is environment-style keys) at a secret path in one
 * environment of a project. Auth is a token, or a universal-auth
 * (machine identity) login with clientid/clientsecret. */
export function infisicalprovider(options?: {
  addr?: string
  token?: string
  clientid?: string
  clientsecret?: string
  project?: string
  environment?: string
  path?: string
}): Provider {
  const opts = options || {}

  // A configured token is kept forever; a universal-auth token carries
  // expiresIn and is renewed shortly before it runs out.
  let livetoken: string | undefined
  let renewat = Infinity

  const login = async (addr: string): Promise<string> => {
    if (opts.token) {
      return opts.token
    }

    if (!opts.clientid || !opts.clientsecret) {
      throw new SekretoError('sekreto: infisical: no token and no client credentials')
    }

    const res = await fetchjson(
      'POST',
      addr + '/api/v1/auth/universal-auth/login',
      { 'content-type': 'application/json' },
      JSON.stringify({ clientId: opts.clientid, clientSecret: opts.clientsecret }),
    )

    const got = res.body && res.body.accessToken
    if (200 !== res.status || !got) {
      throw new SekretoError('sekreto: infisical login failed: ' + res.status)
    }

    const expires = Number(res.body.expiresIn)
    renewat = 0 < expires ? Date.now() + Math.max(expires - 60, 1) * 1000 : Infinity

    return String(got)
  }

  return {
    lookup: async (name: string) => {
      const addr = (opts.addr || 'https://app.infisical.com').replace(/\/$/, '')
      checkaddr(addr)

      const project = opts.project || ''
      const environment = opts.environment || ''
      if ('' === project || '' === environment) {
        throw new SekretoError('sekreto: infisical: no project/environment')
      }

      if (undefined === livetoken || Date.now() >= renewat) {
        livetoken = await login(addr)
      }

      const url =
        addr +
        '/api/v3/secrets/raw/' +
        envkey(name) +
        '?workspaceId=' +
        encodeURIComponent(project) +
        '&environment=' +
        encodeURIComponent(environment) +
        '&secretPath=' +
        encodeURIComponent(opts.path || '/')

      const res = await fetchjson('GET', url, { authorization: 'Bearer ' + livetoken })

      if (404 === res.status) {
        return undefined
      }

      if (200 !== res.status) {
        throw new SekretoError('sekreto: infisical error: ' + res.status)
      }

      const value = res.body && res.body.secret && res.body.secret.secretValue
      return undefined === value || null === value ? undefined : String(value)
    },
    describe: () => 'infisical:' + (opts.project || '') + '/' + (opts.environment || ''),
  }
}


/** The plugin. Needs HTTPS. */
export const infisical = providerplugin('infisical', (spec: ProviderSpec) =>
  infisicalprovider(spec))
